src/blackdoor/cqbe/node/Updater.java

```diff
@@ -1,225 +1,244 @@
+						DBP.printdemoln("Adding " + node + " to table from updater");
+					Node.getAddressTable().add(node);
+					forgive(node);
+				}else{
+					strike(node);
+				}
 			} catch (RPCException e) {
-				DBP.printerrorln(e);
+				// TODO determine if exception was somehow not node's fault, else strike node
+				DBP.printException(e);
 			}
-			for(L3Address c : candidates.values()){
-				if(Node.getAddressTable().contains(c))
-					continue;
-				DBP.printdemoln("Adding " + c + " to address table from updater");
-				toAdd.add(c);
+		}
+		*/
+		//poll for new storage values
+		for(L3Address neighbor : Node.getAddressTable()){
+			try {
+				List<Address> keys = Router.getIndex(neighbor, 1);
+				for(Address key : keys){
+					if (!Node.getStorageController().containsKey(key)) {
+						byte[] value = Router.getValue(neighbor, key);
+						Node.getStorageController().put(
+								new CASFileAddress(Node.getStorageController()
+										.getDomain(), value));
+					}
+				}
+				// too much forgiving: forgive(neighbor);
+			} catch (RPCException e) {
+				DBP.printException(e);
+				if(e.getRPCError().equals(JSONRPCError.INVALID_RESULT))
+					strike(neighbor);
+			} catch (IOException e) {
+				//TODO
+				DBP.printException(e);
 			}
-		return toAdd;
-	}
-	
-	
-	public void updateStorage(){
+		}
+		
+		//update storage controller
+		Node.getStorageController().garbageCollectReferences();
 		try {
-			if(Node.getAddressTable().size() > 0)
-				Node.getStorageController().deleteThirdBucket();
+			Node.getStorageController().deleteThirdBucket();
 		} catch (IOException e) {
-			// TODO Auto-generated catch block
+			// TODO probably shouldn't be throwing this, should change in StorageController
 			e.printStackTrace();
 		}
-		Node.getStorageController().garbageCollectReferences();
-		
 	}
 	
-	
-	
-	
+	private static class AddressUpdateThread implements Runnable{
+		
+		BlockingQueue<L3Address> q;
+		Set<L3Address> everyone;
+		boolean seek;
+		Updater ref;
+		
+		public AddressUpdateThread(BlockingQueue<L3Address> q, Set<L3Address> everyone, boolean seek, Updater ref) {
+			this.q = q;
+			this.everyone = everyone;
+			this.seek = seek;
+			this.ref = ref;
+		}
+
+		@Override
+		public void run() {
+			for(L3Address node = q.poll(); node != null; node = q.poll()){
+				if(seek){
+					try {
+						AddressTable nn = Router.primitiveLookup(node, Node.getOverlayAddress());
+						everyone.addAll(nn.values());
+					} catch (IOException e) {
+						// TODO Auto-generated catch block
+						DBP.printException(e);
+					} catch (RPCException e) {
+						// TODO Auto-generated catch block
+						e.printStackTrace();
+					}
+				}else{
+					try {
+						if(Router.ping(node)){
+							if(!Node.getAddressTable().contains(node))
+								DBP.printdemoln("Adding " + node.l3ToString() + " to table from updater");
+							Node.getAddressTable().add(node);
+							ref.forgive(node);
+						}else{
+							ref.strike(node);
```

Comment:
why don't you combine ping and lookup by sending a lookup only to the nodes that respond to the ping